pkg/middleware/auth.go

```diff
@@ -97,6 +100,65 @@ func CanAdminPlugins(cfg *setting.Cfg, accessControl ac.AccessControl) func(c *c
 	}
 }
 
+func RoleAppPluginAuthAndSignedIn(accessControl ac.AccessControl, ps pluginstore.Store, features featuremgmt.FeatureToggles, logger log.Logger) func(c *contextmodel.ReqContext) {
+	return func(c *contextmodel.ReqContext) {
+		if !c.IsSignedIn {
+			notAuthorized(c)
+			return
+		}
+
+		pluginID := web.Params(c.Req)[":id"]
+		p, exists := ps.Plugin(c.Req.Context(), pluginID)
+		if !exists {
+			c.JsonApiErr(http.StatusNotFound, "Plugin not found", nil)
+			return
+		}
+
+		found := false
+		allowed := false
+		path := normalizeIncludePath(c.Req.URL.Path)
+
+		hasAccess := ac.HasAccess(accessControl, c)
+		for _, i := range p.Includes {
+			if i.Type != "page" {
+				continue
+			}
+
+			if normalizeIncludePath(i.Path) == path {
+				found = true
+
+				useRBAC := features.IsEnabledGlobally(featuremgmt.FlagAccessControlOnCall) && i.RequiresRBACAction()
+				if useRBAC && !hasAccess(ac.EvalPermission(i.Action)) {
+					logger.Debug("plugin include is covered by RBAC, user doesn't have access",
+						"plugin", pluginID,
+						"include", i.Name)
+					allowed = false
+					break
+				} else if !useRBAC && !c.HasUserRole(i.Role) {
+					allowed = false
+					break
+				}
+				allowed = true
+			}
+		}
+
+		if !found {
+			// This isn't an API request, so we probably shouldn't return a JSON error.
+			c.JsonApiErr(http.StatusNotFound, "Plugin page not found", nil)
+			return
+		}
+
+		if !allowed {
+			accessForbidden(c)
+			return
+		}
+	}
+}
+
+func normalizeIncludePath(p string) string {
+	return strings.TrimPrefix(p, "/")
```

Comment:
I think we can protect against path traversal like this:
```suggestion
	return strings.TrimPrefix(filepath.Clean(p), "/")
```